Add tests for user handlers in server_test.go

HarlamovBuldog opened the pull request.

HarlamovBuldog committed:
Add tests for user handlers in server_test.go

pisarukv commented on internal/server/server_test.go:
```diff
@@ -0,0 +1,82 @@
+
+	actualCode := w.Result().StatusCode
+	assert := assert.New(t)
+	assert.Equal(http.StatusOK, actualCode, "The two http codes should be the same")
```
so, why do you want to use assert, but not require?

HarlamovBuldog replied:
Does it make much difference? I mean from documentation: https://godoc.org/github.com/stretchr/testify

assert | Package assert provides a set of comprehensive testing tools for use with the normal Go testing system.

require | Package require implements the same assertions as the `assert` package but stops test execution when a test fails.

I mean in case when only 1 assert call present there is no difference. But in case u mention, if  actual code doesn't match expected, i guess i still need to decode body to get full picture. If i will use require test will fail before check. That's the point i guess.

HarlamovBuldog committed:
Delete occurrences of assert in server_test.go

Both "assert" and "require" utils are from pkg "github.com/stretchr/testify.
But they have very similar behavior. So there is no point to use them both.
So we get rid of "assert" and change all its occurences to "require" call.

HarlamovBuldog committed:
Rename server_test.go to user_test.go

HarlamovBuldog committed:
Add storage mock.

Code generated by MockGen lib to mock Service interface from storage package.
We use this solution for proper testing handlers.

HarlamovBuldog committed:
Code refactoring.

Change method inside requests to POST in tests for CreateNewUser handler.
Move declaring NewSever to place where it's actually starting to be used.
Remove unnecessary declaring of "require" variable.

HarlamovBuldog committed:
Add tests for GetUserInfo and RemoveUser handlers

Request variables are set with help of SetURLVars func of gorilla/mux lib.
Cause mux.Vars can't read URL vars from httptest.NewRequest.

HarlamovBuldog committed:
Add tests for Take- and AddUserBonusPoints handlers.

HarlamovBuldog committed:
Use Sprintf to create expectedURLPath in user_test.go

Problem here is that we are not cheking request method and url path at all.
Also we are not parsing URL path string properly.
For now we are setting request variables manually.
Need to parse URL path and get those vars.

HarlamovBuldog commented:
Having hard times with checking request URL path and method. Cause for now inside tests these two are not checked at all.

HarlamovBuldog commented:
Tried to do smth like this. But not sure if it's the best solution.

```
b := bytes.NewBuffer(enc)
req := httptest.NewRequest("POST", "/user", b)

actualURLPath := req.URL.Path
require.Equal("/user", actualURLPath, "The two URL pathes should be the same")

actualReqMethod := req.Method
require.Equal("POST", actualReqMethod, "The two request methods should be the same")
```

pisarukv commented on internal/server/user_test.go:
```diff
@@ -0,0 +1,426 @@
+	defer ctrl.Finish()
+
+	mock := storage.NewMockService(ctrl)
+	mock.EXPECT().AddUser(gomock.Any(), gomock.Eq("Gennadiy")).Times(1).Return("code_str", nil)
```
code_str to const

HarlamovBuldog committed:
Move "user_code" to string variable in user_test.go.
